### PR TITLE
Remove JSCS as it has merged with ESLint

### PR DESCRIPTION
**Pull request by hekike**
## Pre-Submission Checklist

- [x] Opened an issue discussing these changes before opening the PR
- [x] Ran the linter and tests via `make prepush`

## Issues

Closes:

- https://github.com/restify/node-restify/issues/1315

JSCS has merged with ESLint: 
https://medium.com/@markelog/jscs-end-of-the-line-bc9bf0b3fdb2

# Changes

- Remove JSCS
- Migrate JSCS rules to ESLint
- Enable uncommented stricter ESLint rules
- Separate make command for linting and styling
- Auto-fix indentation issues
- Fix `self` assignment issues (`self` "keyword" was used for other than `this`)

### CODE REVIEW

**Comment by retrohacker** [comment]
Yes! Please! :heart:

**Comment by hekike** [comment]
@yunong if we separate the two things into multiple `.eslintrc` files the IDE cannot make live linting based on the `.eslintrc`. Which would be a huge comfort killer.

What would be the benefit of keeping them separate?

**Comment by hekike** [comment]
For the record: 
We discussed the benefits of having a separate `lint` and `style` commands and come up with a solution that is based on the `.eslintrc.js` file and uses environment variables to disable a specific group of rules in certain cases. This concept also allows us to have one ESLint config file that works well with most of the popular IDEs.

"Unfortunately" ESLint's indentation checker is much stricter that caused a massive and evil (really: `+666 −666`) line changes. I moved it to a separate commit.

I think this PR is ready for merging in the case you agree with me.

cc @yunong @retrohacker

**Comment by retrohacker** [comment]
> I'm generally happy do this, however I'd like to still keep the .eslintrc files for style and lint separate 

> We discussed the benefits of having a separate lint and style 

Can I get some more context on this @yunong? What is the value add?

**Comment by retrohacker** [comment]
I'm :+1: on merging this either way since it unblocks @hekike's work on other PRs, and I would just like to see JSCS leave the codebase.

I'd like to at least revisit the decision to differentiate between linting and style rules via custom tooling (is there a distinction? Are there rules that fit both? Is the cognitive overhead of separating them worth it? What value do we loose by combining them? What does supporting this custom tooling moving forward look like, and is that worth it vs. a plain old .eslintrc file? Does this increase a barrier to entry for non-core members contributing linting rules? Does this translate easily to other projects we maintain?) but we can move that to another Issue/PR if @yunong is :+1: on landing this.

File: lib/plugins/cpuUsageThrottle.js [hunk 1]
```diff
@@ -86,39 +86,41 @@ function cpuUsageThrottlePlugin (opts) {
...
     // current point in time based on how much over our limit we are. This is
     // updated on an interval by updateReject().
-    self._reject = 0;
+    plugin._reject = 0;
 
-    // self._timeout keeps track of the current handle for the setTimeout we
+    // plugin._timeout keeps track of the current handle for the setTimeout we
     // use to gather CPU load averages, this allows us to cancel the timeout
     // when shutting down restify.
-    self._timeout = null;
```
Comment (yunong) [inline lib/plugins/cpuUsageThrottle.js]: Why is it changing all the `self`s to `plugin`?

File: tools/mk/Makefile.targ [hunk 2]
```diff
@@ -122,16 +122,12 @@ check-bash: $(BASH_FILES:%=%.bashchk) $(BASH_FILES:%=%.bashstyle)
...
-check-jscs::
-	$(JSCS) $(JS_FILES)
-
 .PHONY: check-eslint
 check-eslint::
 	$(ESLINT) $(JS_FILES)
 
 .PHONY: check
-check: check-eslint check-jscs check-bash
+check: check-eslint check-bash
```
Comment (yunong) [inline tools/mk/Makefile.targ]: Did you get a chance to add back the style target?

File: lib/plugins/cpuUsageThrottle.js (hunk 1 above)
Comment (hekike) [reply lib/plugins/cpuUsageThrottle.js]: `consistent-this` rule forces to use `self` for `this`

Comment (hekike) [reply lib/plugins/cpuUsageThrottle.js]: Self is not the `this` here.

File: tools/mk/Makefile.targ (hunk 2 above)
Comment (hekike) [reply tools/mk/Makefile.targ]: Without the environment variables, it runs all of the rules both lint and style. Do you want to separate hem in the `Makefile.targ` like:  `check-lint` and `check-style`? I would keep `check-eslint` for the default behavior.

Comment (hekike) [reply tools/mk/Makefile.targ]: Probably I should rename these: https://github.com/restify/node-restify/pull/1527/files#diff-b67911656ef5d18c4ae36cb6741b7965R52

Comment (hekike) [reply tools/mk/Makefile.targ]: Could be something like this:

```
.PHONY: check-eslint
check-eslint::
	$(ESLINT) $(JS_FILES)

.PHONY: check-lint
check-lint::
	NO_STYLE=true $(ESLINT) $(JS_FILES)

.PHONY: check-style
check-style::
	NO_LINT=true $(ESLINT) $(JS_FILES)

.PHONY: check
check: check-lint check-style check-bash
	@echo check ok
```

Comment (yunong) [reply tools/mk/Makefile.targ]: Yes, if we can keep two separate targets it'll keep the makefile as it was, but all on eslint which would be awesome.

Comment (hekike) [reply tools/mk/Makefile.targ]: Fixed

**Comment by retrohacker** [comment]
Leaving a comment here for future Will with the context we talked about offline:

We have two classes of rules that we would like to keep separated: correctness and style. We have an opinionated set of rules that we believe help make a Node.js application "correct" by removing certain classes of errors. These rules fall under "linting." We then have an opinionated set of styles like indentation that don't necessarily improve the correctness of the code but help improve it's readability.

By keeping these separate, we can easily copy the "correctness" rules to our other projects that we maintain while allowing flexibility for different styles in those code bases.

Keeping this delineation right now didn't introduce significant complexity. If we find this breaks forwards compatibility in the future, we may revisit this decision.

**Pull request merged**